src/test/ui/iterators/into-iter-on-arrays-lint.rs

```diff
@@ -0,0 +1,22 @@
+// run-pass
+
+// If this test fails for you, you might have moved the definition of
+// `core::iter::IntoIterator`. If so, please adjust the path in
+// `src/librustc_lint/array_into_iter.rs`.
+
+fn main() {
+    // Expressions that should trigger the lint
+    let array = [1, 2];
+    array.into_iter();
+    (&array).into_iter();
+
+    [1, 2].into_iter();
+    (&[1, 2]).into_iter();
```

Comment:
It's not obvious to me that this one is desirable to lint on.  Would it be feasible to detect the auto-ref (like in 13) and only trigger the lint when it was involved?  (AFAICT this one wouldn't actually be broken by the IntoIterator for arrays.)

That said, I'm fine with linting on this, because it's strange code anyway and I'd rather people change it.

Reply:
+1, this shouldn't be linted against for the same reason `for x in &[1, 2] { ... }` shouldn't be linted against.

Reply:
How would one check for this? So in HIR in both cases, the first method argument has the type `Ref(Array(..))`. I guess one can get from HIR node to AST node, but even if: the AST is purely syntactical right? I.e. I don't have type information, right? I basically want to know: was autoref applied to the `self` receiver, right? Could someone give me a hint how one could find that out?

PS: I basically looked at how the clippy lint is implemented and used that check. So the clippy lint applies to `(&[1, 2]).into_iter()`, too, I think.